keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,421 @@
+		klog.V(3).Infof("Volume %q is already being deleted", volume.Name)
+		return "", nil
+	}
+    // Remaining code below skipped for brevity...
+    // Please refer pkg/controller/volume/persistentvolume/pv_controller.go in kubernetes/kubernetes for the full code.
+```
+11. Meanwhile, the external-provisioner checks if there is a `deletionTimestamp` on the PV, if so, it assumes that its
+in a transitory state and returns false for the `shouldDelete` check.
+```go
+// shouldDelete returns whether a volume should have its backing volume
+// deleted, i.e. whether a Delete is "desired"
+func (ctrl *ProvisionController) shouldDelete(ctx context.Context, volume *v1.PersistentVolume) bool {
+	if deletionGuard, ok := ctrl.provisioner.(DeletionGuard); ok {
+		if !deletionGuard.ShouldDelete(ctx, volume) {
+			return false
+		}
+	}
+
+	if ctrl.addFinalizer && !ctrl.checkFinalizer(volume, finalizerPV) && volume.ObjectMeta.DeletionTimestamp != nil {
+		return false
+	} else if volume.ObjectMeta.DeletionTimestamp != nil { //<========== DeletionTimestamp is set on the PV since it's deleted first.
+		return false
+	}
+    // Remaining code below skipped for brevity...
+    // Please refer sig-storage-lib-external-provisioner/controller/controller.go for the full code.
+```
+
+The main approach in fixing the issue involves introducing new finalizer `kubernetes.io/pv-deletion-protection` on the
+Persistent Volume when it is observed that it has `DeletionTimestamp` set but still associated with a valid claim.
+Adding the finalizer prevents the PV to be removed from the API server.
+
+When the PVC is deleted, the PV is moved into `Released` state and following checks are made:
+
+1. Plugin:
+If the volume has the finalizer `kubernetes.io/pv-deletion-protection`, then, `DeletionTimestamp` checks can be ignored.
+
+3. CSI driver
+
+If when processing the PV update it is observed that it has `kubernetes.io/pv-deletion-protection` finalizer and
+`DeletionTimestamp` set, then the volume deletion request is forwarded to the driver(provided other pre-defined 
+conditions are met).
+
+
+
+### Risks and Mitigations
+
+The primary risk is volume deletion to a user that expect the current behavior, i.e, they do not expect the volume to be
+deleted from the underlying storage infrastructure when for a bound PV-PVC pair, the PV is deleted followed by deleting
+the PVC. This can be mitigated by initially introducing the fix behind a feature gate, followed by explicitly deprecating
+the current behavior.
+
+## Design Details
+
+When processing the PV update in `kube-controller-manager`, if it's observed that the PV is still bound to a valid claim,
+and also has `DeletionTimestamp` set, then a finalizer is added to the PV.
+
+```go
+// PvDeletionProtectionFinalizer finalizer is added to the PV when it is observed that the PV
+// has a deletion timestamp but still associated with a valid claim.
+PvDeletionProtectionFinalizer = "kubernetes.io/pv-deletion-protection"
```

Comment:
PV is an abbreviation, therefore it must be upper case, `PVDeletionProtectionFinalizer`